smart_open/s3.py

```diff
@@ -279,47 +279,92 @@ class _SeekableRawReader(object):
     This class is internal to the S3 submodule.
     """
 
-    def __init__(self, s3_object, content_length, version_id=None, object_kwargs=None):
+    def __init__(self, s3_object, version_id=None, object_kwargs=None):
         self._object = s3_object
-        self._content_length = content_length
+        self._content_length = None
         self._version_id = version_id
         self._position = 0
         self._body = None
         self._object_kwargs = object_kwargs if object_kwargs else {}
 
-    def seek(self, position):
-        """Seek to the specified position (byte offset) in the S3 key.
+    def seek(self, offset, whence=constants.WHENCE_START):
+        """Seek to the specified position.
```

Comment:
The `whence` logic and `range_string` logic both have to move into this seek() so we can handle the case of reading bytes from the end of a file:

```
with open("s3://my-bucket/my-object") as f:
    f.seek(-1000, constants.WHENCE_END)
    f.read()
```

The above only requires one `GetObject` if we call it with `Range: bytes=-1000`.